### PR TITLE
Add SetTag endpoint

### PR DESCRIPTION
**Pull request by jescalada**
PR for adding the `SetTag` endpoint. I also added the `DeleteTag` endpoint from #8.

There are various new tests, all of which are passing except one:
- PASSING: `test_set_tag_with_empty_string_as_value[sqlalchemy]`
- PASSING: `test_set_tag_validation[sqlalchemy]`
- PASSING: `test_set_tag_truncate_too_long_tag`
- PASSING: `test_update_run_name`
- PASSING: `test_update_run_name_without_changing_status[sqlalchemy]`
- **FAILING**: `test_set_tag`

The reason why `test_set_tag` is failing, is that it uses `monkeypatch` to set an env, which doesn't flow to the Go side in our implementation. I have temporarily disabled that specific test by adding it to `conftest.py` until we can find a solution for similar tests.

I also added a `test:PythonSpecific` command for testing a single endpoint at a time.

### CODE REVIEW

**Comment by jescalada** [comment]
Looks like I messed up the line endings, will fix in a sec!

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -59,8 +59,8 @@
 	// "forwardPorts": [5432],
 
 	// Use 'postCreateCommand' to run commands after the container is created.
-	"postCreateCommand": ".devcontainer/postCreate.sh"
+	"postCreateCommand": ".devcontainer/postCreate.sh",
 
 	// Uncomment to connect as root instead. More info: https://aka.ms/dev-containers-non-root.
-	// "remoteUser": "root"
+	"remoteUser": "root"
```
Comment (nojaf) [inline .devcontainer/devcontainer.json:65]: Why did you need this?

File: magefiles/tests.go [hunk 2]
```diff
@@ -64,6 +64,33 @@ func (Test) Python() error {
 	return nil
 }
 
+// Run specific Python test against the Go backend.
+func (Test) PythonSpecific(testName string) error {
```
Comment (nojaf) [inline magefiles/tests.go:68]: Try and refactor this from the existing code, you should be able to extract.

File: pkg/tracking/service/tags.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/mlflow/mlflow-go/pkg/contract"
+	"github.com/mlflow/mlflow-go/pkg/protos"
+)
+
+func (ts TrackingService) SetTag(ctx context.Context, input *protos.SetTag) (*protos.SetTag_Response, *contract.Error) {
+	// Print input
```
Comment (nojaf) [inline pkg/tracking/service/tags.go:12]: Either use logger via the context, or remove temporary print.

File: pkg/tracking/store/sql/tags.go [hunk 4]
```diff
@@ -80,3 +81,182 @@ func (s TrackingSQLStore) setTagsWithTransaction(
...
 	return nil
 }
+
+const (
+	maxEntityKeyLength = 250
+	maxTagValueLength  = 8000
+)
+
+// Helper function to validate the tag key and value
+func validateTag(key, value string) *contract.Error {
```
Comment (nojaf) [inline pkg/tracking/store/sql/tags.go:91]: This should not at all happen here.
The validation code happens in the `generate/validations.go`.
You have put `required` in there, this will check for empty strings.

File: pkg/tracking/store/sql/tags.go [hunk 5]
```diff
@@ -80,3 +81,182 @@ func (s TrackingSQLStore) setTagsWithTransaction(
...
+			"Missing value for required parameter 'key'",
+		)
+	}
+	if len(key) > maxEntityKeyLength {
+		return contract.NewError(
+			protos.ErrorCode_INVALID_PARAMETER_VALUE,
+			fmt.Sprintf("Tag key '%s' had length %d, which exceeded length limit of %d", key, len(key), maxEntityKeyLength),
+		)
+	}
+	if len(value) > maxTagValueLength {
```
Comment (nojaf) [inline pkg/tracking/store/sql/tags.go:104]: Needs a `max: ` annotation on the validation side.

File: pkg/tracking/store/sql/tags.go [hunk 6]
```diff
@@ -80,3 +81,182 @@ func (s TrackingSQLStore) setTagsWithTransaction(
...
+		return contract.NewError(
+			protos.ErrorCode_INVALID_PARAMETER_VALUE,
+			fmt.Sprintf("Tag value exceeded length limit of %d characters", maxTagValueLength),
+		)
+	}
+	// TODO: Check if this is the correct way to prevent invalid values
+	if _, err := strconv.ParseFloat(value, 64); err == nil {
```
Comment (nojaf) [inline pkg/tracking/store/sql/tags.go:111]: Where is this coming from? A float as tag value is weird, but I would not say it is invalid perse.

File: pkg/tracking/store/sql/tags.go [hunk 7]
```diff
@@ -80,3 +81,182 @@ func (s TrackingSQLStore) setTagsWithTransaction(
...
+) *contract.Error {
+	if runID == "" {
+		return contract.NewError(
+			protos.ErrorCode_INVALID_PARAMETER_VALUE,
+			"RunID cannot be empty",
+		)
+	}
+
+	// If the runID can be parsed as a number, it should throw an error
+	if _, err := strconv.ParseFloat(runID, 64); err == nil {
```
Comment (nojaf) [inline pkg/tracking/store/sql/tags.go:131]: This again is weird, where is this coming from? Feel very unnecessary.

**Review by nojaf — COMMENTED** [review]
Hi @jescalada, thanks for this contribution!
There are some interesting changes in here which I did not expect.
It would be great to jump on a call sometime and further discuss.

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (jescalada) [reply .devcontainer/devcontainer.json:65]: When setting up the repo, the first issue I had was that apparently I didn't have enough permissions on the devcontainer to run the server. This might be due to using Dev Containers on Windows... I forgot to revert before committing!

File: pkg/tracking/store/sql/tags.go (hunk 4 above)
Comment (jescalada) [reply pkg/tracking/store/sql/tags.go:91]: When I wrote this code, I was trying to fix the failing tests which seemed related to the validation. Upon investigating further, the test result doesn't change when removing the `validateTag` function.

File: pkg/tracking/store/sql/tags.go (hunk 7 above)
Comment (jescalada) [reply pkg/tracking/store/sql/tags.go:131]: This is also due to some of the weird edge cases that I thought were failing. This is what the previous Python tests looked like:

```python
def assert_bad_request(payload, expected_error_message):
      response = _send_rest_tracking_post_request(
          mlflow_client.tracking_uri,
          "/api/2.0/mlflow/runs/set-tag",
          payload,
      )
      assert response.status_code == 400
      assert expected_error_message in response.text

assert_bad_request(
  {
      "run_id": 31,
      "key": "tag",
      "value": 41,
  },
  "Invalid value 31 for parameter 'run_id' supplied",
)
assert_bad_request(
  {
      "run_id": run_id,
      "key": "param",
      "value": 41,
  },
  "Invalid value 41 for parameter 'value' supplied",
)
assert_bad_request(
  {
      "run_id": run_id,
      # Missing key
      "value": "value",
  },
  "Missing value for required parameter 'key'",
)

response = _send_rest_tracking_post_request(
  mlflow_client.tracking_uri,
  "/api/2.0/mlflow/runs/set-tag",
  {
      "run_uuid": run_id,
      "key": "key",
      "value": "value",
  },
)
```

Comment (jescalada) [reply pkg/tracking/store/sql/tags.go:131]: It seems that IDs are expected to be in the UUID form rather than simple integer form

File: pkg/tracking/store/sql/tags.go (hunk 6 above)
Comment (jescalada) [reply pkg/tracking/store/sql/tags.go:111]: This is also due to my misunderstanding in how validations happen! The comment below shows my reasoning for adding the validation.

File: .devcontainer/devcontainer.json [hunk 8]
```diff
@@ -59,7 +59,7 @@
 	// "forwardPorts": [5432],
 
 	// Use 'postCreateCommand' to run commands after the container is created.
-	"postCreateCommand": ".devcontainer/postCreate.sh"
+	"postCreateCommand": ".devcontainer/postCreate.sh",
```
Comment (nojaf) [inline .devcontainer/devcontainer.json:62]: Please revert as well

File: magefiles/tests.go [hunk 9]
```diff
@@ -28,8 +28,7 @@ func cleanUpMemoryFile() error {
 	return nil
 }
 
-// Run mlflow Python tests against the Go backend.
-func (Test) Python() error {
+func RunPythonTests(testFiles []string, testName string) error {
```
Comment (nojaf) [inline magefiles/tests.go:31]: This should be lowercased so that the function is private.

File: pkg/protos/scalapb/scalapb.pb.go [hunk 10]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
 // 	protoc-gen-go v1.33.0
-// 	protoc        v5.26.0
+// 	protoc        v3.21.12
```
Comment (nojaf) [inline pkg/protos/scalapb/scalapb.pb.go:4]: This is weird, did you use the dev container to run `mage generate`?
I would not have expected this to change.

Comment (jescalada) [reply pkg/protos/scalapb/scalapb.pb.go:4]: Yes, as far as I recall! 🤔

**Comment by jescalada** [comment]
I managed to get the `test_set_tag_validation` test to pass, however, I did this via a manual check in `/pkg/tracking/service/tags.go`. I also added the custom validator for the `runId`, however I am not sure how to integrate this with the existing code.

Is this meant to be used in our own Go tests? I noticed that for the `LogBatch` we're doing custom tests in `/pkg/validation/validation_test.go`.

File: pkg/validation/validation.go [hunk 11]
```diff
@@ -107,6 +107,15 @@ func validateLogBatchLimits(structLevel validator.StructLevel) {
 	}
 }
 
+// SetTag must have either a run_id or a run_uuid present.
+func validateSetTagRunIDExists(structLevel validator.StructLevel) {
```
Comment (nojaf) [inline pkg/validation/validation.go:111]: This looks good, but does that produce the correct error message?

Comment (jescalada) [reply pkg/validation/validation.go:111]: I believe the error message isn't getting checked anywhere. Should we perhaps extend the Python tests to make sure this is in order? Or is this something we handle in the Go tests?

I did notice that unlike the other tests (for validation), the run_uuid/run_id test doesn't actually care about an error message. I'm guessing that the only undesirable case is when we have neither of them (and we would just throw an empty `run_id` error). However, there doesn't seem to be a test for this.

File: magefiles/generate/validations.go [hunk 12]
```diff
@@ -25,4 +25,8 @@ var validations = map[string]string{
 	"LogMetric_Key":                      "required",
 	"LogMetric_Value":                    "required",
 	"LogMetric_Timestamp":                "required",
+	"SetTag_Key":                         "required,max=1000,validMetricParamOrTagName,pathIsUnique",
+	"SetTag_Value":                       "required,max=8000",
```
Comment (nojaf) [inline magefiles/generate/validations.go:29]: Needs to be `truncate=8000` instead.

File: pkg/tracking/store/sql/tags.go [hunk 13]
```diff
@@ -80,3 +80,136 @@ func (s TrackingSQLStore) setTagsWithTransaction(
 
 	return nil
 }
+
+func (s TrackingSQLStore) SetTag(
+	ctx context.Context, runID, key, value string,
+) *contract.Error {
+	if runID == "" {
```
Comment (nojaf) [inline pkg/tracking/store/sql/tags.go:87]: Go validations should already have covered this.

File: pkg/tracking/store/sql/tags.go [hunk 14]
```diff
@@ -80,3 +80,136 @@ func (s TrackingSQLStore) setTagsWithTransaction(
 
 	return nil
 }
+
+func (s TrackingSQLStore) SetTag(
```
Comment (nojaf) [inline pkg/tracking/store/sql/tags.go:84]: Wait, I'm confused now, did you try and port

```py
    def set_tag(self, run_id, tag):
        """
        Set a tag on a run.

        Args:
            run_id: String ID of the run.
            tag: RunTag instance to log.
        """
        with self.ManagedSessionMaker() as session:
            tag = _validate_tag(tag.key, tag.value)
            run = self._get_run(run_uuid=run_id, session=session)
            self._check_run_is_active(run)
            if tag.key == MLFLOW_RUN_NAME:
                run_status = RunStatus.from_string(run.status)
                self.update_run_info(run_id, run_status, run.end_time, tag.value)
            else:
                # NB: Updating the run_info will set the tag. No need to do it twice.
                session.merge(SqlTag(run_uuid=run_id, key=tag.key, value=tag.value))
```

`FAILED .mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py::test_update_run_name` is failing because `tag.key == MLFLOW_RUN_NAME` is not ported.

**Review by nojaf — COMMENTED** [review]
`FAILED .mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py::test_set_tag` is failing because the test uses `monkeypatch.setenv("MLFLOW_TRUNCATE_LONG_VALUES", "true")`.

@dsuhinin can you give some advice on what should be done in this case.

**Comment by nojaf** [comment]
Oh and please rebase on latest main.

File: pkg/tracking/store/sql/tags.go (hunk 13 above)
Comment (jescalada) [reply pkg/tracking/store/sql/tags.go:87]: When I remove that check, the tests seem to fail... How can I make sure that the custom validator I wrote is set up and working as expected? 🤔

File: pkg/tracking/store/sql/tags.go (hunk 14 above)
Comment (jescalada) [reply pkg/tracking/store/sql/tags.go:84]: Oh, I thought that part had not been implemented in other endpoints, now I realized that the `MLFLOW_RUN_NAME` value was aliased to `TagRunName` in the Go code!

I also just realized that the `test_set_tag` and `test_update_run_name` were failing all along. I'll take a shot at implementing the missing logic now.

File: pkg/tracking/store/sql/tags.go (hunk 13 above)
Comment (nojaf) [reply pkg/tracking/store/sql/tags.go:87]: You could easily first manually check it using `mage dev`. Just as a sanity it is kicking in.

File: .devcontainer/postCreate.sh
Comment (nojaf) [inline .devcontainer/postCreate.sh:1]: Please revert this change.

File: pkg/tracking/store/sql/tags.go (hunk 13 above)
Comment (nojaf) [reply pkg/tracking/store/sql/tags.go:150]: This should still not be checked here, your validation covered this.

File: pkg/tracking/store/sql/runs_internal_test.go [hunk 15]
```diff
@@ -1,4 +1,3 @@
-//nolint:ireturn
```
Comment (nojaf) [inline pkg/tracking/store/sql/runs_internal_test.go:1]: Please revert this as well. This must be the reason the linter is triggered on CI.

File: .devcontainer/postCreate.sh
Comment (jescalada) [reply .devcontainer/postCreate.sh:1]: I have a feeling that the file was previously CRLF and the commit where I turned everything into LF might have converted it (thus causing this mystery empty change).

**Review by nojaf — APPROVED** [review]
Thanks!

File: magefiles/tests.go [hunk 16]
```diff
@@ -45,25 +44,50 @@ func (Test) Python() error {
...
+	return runPythonTests([]string{
+		".mlflow.repo/tests/tracking/test_rest_tracking.py",
+		".mlflow.repo/tests/tracking/test_model_registry.py",
+		".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
+		".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py",
+	}, "")
+}
+
+// Run specific Python test against the Go backend.
+func (Test) PythonSpecific(testName string) error {
```
[the remark this answers is not in the document]
Comment (jescalada) [reply magefiles/tests.go:82]: I added all files so that it wouldn't miss some of the tests in other files (at first I was not testing comprehensively because I was only looking at the `test_sqlalchemy_store.py`). It matches multiple files so that you can test anything related to your endpoint in one go.

[the remark this answers is not in the document]
Comment (nojaf) [reply magefiles/tests.go:82]: @dsuhinin a mage target cannot have optional arguments.
Maybe we should change the target slightly to take a single string argument and we can pass all pytest arguments in there. When you want to run a specific test you can then do `mage pythonSpecific "filename.py::test-name -vv"`

File: tests/override_test_sqlalchemy_store.py [hunk 17]
```diff
@@ -13,6 +13,10 @@ def test_log_param_max_length_value(store: SqlAlchemyStore, monkeypatch):
     ()
 
 
+def test_set_tag(store: SqlAlchemyStore, monkeypatch):
+    ()
+
+
```
[the remark this answers is not in the document]
Comment (jescalada) [reply tests/override_test_sqlalchemy_store.py:19]: @dsuhinin I just tried implementing this, and it seems to require a lot of test-specific dependencies from the MLflow side. For example, the `_run_factory` function makes use of another helper function `_create_experiments` and so on, which would clutter our tests if we ported them over. And they are private so I don't think we have a choice other than to implement them directly on our `override_test_sqlalchemy_store.py`. I feel that implementing these would clutter that file too much, maybe we could have another file for test utils? 🤔 At any rate, I feel that it would be a pretty large change so we could make a separate issue/PR to sort it out!

File: pkg/tracking/store/sql/tags.go [hunk 18]
```diff
@@ -143,3 +143,109 @@ func (s TrackingSQLStore) DeleteTag(
...
+	ctx context.Context, run models.Run, runID, value string,
+) *contract.Error {
+	runStatus := run.Status.String()
+
+	var endTimePtr *int64
+	if run.EndTime.Valid {
+		endTimePtr = &run.EndTime.Int64
+	}
+
+	if err := s.UpdateRun(ctx, runID, runStatus, endTimePtr, value); err != nil {
```
[the remark this answers is not in the document]
Comment (jescalada) [reply pkg/tracking/store/sql/tags.go:242]: Works without issue! 👍🏼

File: magefiles/tests.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (jescalada) [reply magefiles/tests.go:82]: I have been working on implementing the single string argument testing without much luck. I'm currently working in a separate branch as I was hoping this could get merged in even if the test scripts are not perfect! :)

This I have right now in `tests.go` if you'd like to test it. I'm having issues with the filename matching, right now it keeps executing all the tests instead of just the ones I ask for:
```go
//go:build mage

//nolint:wrapcheck
package main

import (
	"os"

	"github.com/magefile/mage/mg"
	"github.com/magefile/mage/sh"
)

type Test mg.Namespace

func cleanUpMemoryFile() error {
	// Clean up :memory: file
	filename := ":memory:"
	_, err := os.Stat(filename)

	if err == nil {
		// File exists, delete it
		err = os.Remove(filename)
		if err != nil {
			return err
		}
	}

	return nil
}

func runPythonTests(testFiles []string, testName string) error {
	libpath, err := os.MkdirTemp("", "")
	if err != nil {
		return err
	}

	// Remove the Go binary
	defer os.RemoveAll(libpath)
	//nolint:errcheck
	defer cleanUpMemoryFile()

	// Build the Go binary in a temporary directory
	if err := sh.RunV("python", "-m", "mlflow_go.lib", ".", libpath); err != nil {
		return nil
	}

	args := []string{
		"--confcutdir=.",
	}

	// Use testFiles directly if specified, otherwise use defaults
	if len(testFiles) > 0 {
		args = append(args, testFiles...)
	} else {
		args = append(args, ".mlflow.repo/tests/tracking/test_rest_tracking.py",
			".mlflow.repo/tests/tracking/test_model_registry.py",
			".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
			".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py")
	}

	// If testName is provided, append it directly
	if testName != "" {
		args = append(args, testName) // No need for -k if full test path is used
	}

	//  Run the tests (currently just the server ones)
	if err := sh.RunWithV(map[string]string{
		"MLFLOW_GO_LIBRARY_PATH": libpath,
	}, "pytest", args...); err != nil {
		return err
	}

	return nil
}

// Run mlflow Python tests against the Go backend.
func (Test) Python(pytestArgs string) error {
	return runPythonTests([]string{
		".mlflow.repo/tests/tracking/test_rest_tracking.py",
		".mlflow.repo/tests/tracking/test_model_registry.py",
		".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
		".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py",
	}, pytestArgs)
}

// Run specific Python test against the Go backend.
func (Test) PythonSpecific(pytestArgs string) error {
	return runPythonTests([]string{
		".mlflow.repo/tests/tracking/test_rest_tracking.py",
		".mlflow.repo/tests/tracking/test_model_registry.py",
		".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
		".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py",
	}, pytestArgs)
}

// Run the Go unit tests.
func (Test) Unit() error {
	return sh.RunV("go", "test", "./pkg/...")
}

// Run all tests.
func (Test) All() {
	mg.Deps(Test.Unit, func() error {
		return Test{}.Python("")
	})
}
```

[the remark this answers is not in the document]
Comment (nojaf) [reply magefiles/tests.go:76]: I refactored this a little, tl;dr you can pass all the files I think.
It makes more sense to accept an argument like `pytest` would expect.

[the remark this answers is not in the document]
Comment (jescalada) [reply magefiles/tests.go:76]: Ohh, I got it now, thanks @nojaf! Feel free to merge the PR now as I don't have write access.
